pyteal/ast/subroutine.py

```diff
@@ -461,53 +648,93 @@ def evaluateSubroutine(subroutine: SubroutineDefinition) -> SubroutineDeclaratio
-        loaded: Union[ScratchVar, abi.BaseType, Expr]
-        argVar: ScratchVar
+    ) -> tuple[ScratchVar, ScratchVar | abi.BaseType | Expr]:
+        loaded_var: ScratchVar | abi.BaseType | Expr
+        argument_var: ScratchVar
 
         if param in subroutine.by_ref_args:
-            argVar = DynamicScratchVar(TealType.anytype)
-            loaded = argVar
+            argument_var = DynamicScratchVar(TealType.anytype)
+            loaded_var = argument_var
         elif param in subroutine.abi_args:
             internal_abi_var = subroutine.abi_args[param].new_instance()
-            argVar = internal_abi_var.stored_value
-            loaded = internal_abi_var
+            argument_var = internal_abi_var.stored_value
+            loaded_var = internal_abi_var
         else:
-            argVar = ScratchVar(TealType.anytype)
-            loaded = argVar.load()
+            argument_var = ScratchVar(TealType.anytype)
+            loaded_var = argument_var.load()
 
-        return argVar, loaded
+        return argument_var, loaded_var
+
+    if len(subroutine.output_kwarg) > 1:
+        raise TealInputError(
+            f"ABI keyword argument num: {len(subroutine.output_kwarg)}. "
+            f"Exceeding abi output keyword argument max number 1."
+        )
 
     args = subroutine.arguments()
-    argumentVars, loadedArgs = zip(*map(var_n_loaded, args)) if args else ([], [])
+    args = [arg for arg in args if arg not in subroutine.output_kwarg]
+
+    arg_vars: list[ScratchVar] = []
+    loaded_args: list[ScratchVar | Expr | abi.BaseType] = []
+    for arg in args:
+        arg_var, loaded_arg = var_n_loaded(arg)
+        arg_vars.append(arg_var)
+        loaded_args.append(loaded_arg)
+
+    abi_output_kwargs: dict[str, abi.BaseType] = {}
+    output_kwarg_info = OutputKwArgInfo.from_dict(subroutine.output_kwarg)
+    output_carrying_abi: Optional[abi.BaseType] = None
+
+    if output_kwarg_info:
+        output_carrying_abi = output_kwarg_info.abi_type.new_instance()
+        abi_output_kwargs[output_kwarg_info.name] = output_carrying_abi
 
     # Arg usage "B" supplied to build an AST from the user-defined PyTEAL function:
-    subroutineBody = subroutine.implementation(*loadedArgs)
+    subroutine_body = subroutine.implementation(*loaded_args, **abi_output_kwargs)
 
-    if not isinstance(subroutineBody, Expr):
+    if not isinstance(subroutine_body, Expr):
         raise TealInputError(
-            f"Subroutine function does not return a PyTeal expression. Got type {type(subroutineBody)}"
+            f"Subroutine function does not return a PyTeal expression. Got type {type(subroutine_body)}."
+        )
+    # if there is an output keyword argument for ABI, place the storing on the stack
+    if output_carrying_abi:
+        if subroutine_body.has_return():
+            raise TealInputError(
+                "ABI returning subroutine definition should have no return"
```

Comment:
I _think_ you're choosing to error in this case because it would be impossible to add the output ABI value to the stack if the subroutine has a retsub op?

FYI, `has_return()` is only true if _every_ possible codepath in the expression results in a return. So if only some paths end in a return, it would be false, meaning that case isn't addressed here.

Separately, I think it _should_ be possible for users to use the `Return()` expression in a subroutine, since it's a useful feature. But it looks like supporting this might be more difficult than I realized.

Reply:
good point, I was assuming that `has_return()` is for *existence* of return expression in expr, now it seems wrong.

I was going down this path, and append `Return(output.get())` to the subroutine body.

Now it seems we should let users to write `Return(output.get())` themselves in the expr, and we still need to check `has_return` to make sure that the subroutine is well-formed, i.e., all code paths returns.

Does that sound reasonable to you?

Reply:
Think a little bit more of your suggestion, and I realized that one can probably just directly return the underlying representation of ABI value, i.e., `Return(Int(1))`, for an output of `abi.Uint64`.

Feels somewhat hacky, but seems okay to me. Any thoughts on it?

Reply:
ok, seems to be resolving in #328